stan/math/rev/core/scoped_chainablestack.hpp

```diff
@@ -0,0 +1,70 @@
+#ifndef STAN_MATH_REV_CORE_SCOPED_CHAINABLESTACK_HPP
+#define STAN_MATH_REV_CORE_SCOPED_CHAINABLESTACK_HPP
+
+#include <stan/math/rev/core/chainablestack.hpp>
+
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * The AD tape of reverse mode AD is by default stored globally within the
+ * process (or thread). With the ScopedChainableStack class one may execute a
+ * nullary functor with reference to an AD tape which is stored with the
+ * instance of ScopedChainableStack. Example:
+ *
+ * ScopedChainableStack scoped_stack;
+ *
+ * double cgrad = scoped_stack.execute([] {
+ *   var a = 1;
+ *   var b = 4;
+ *   var c = a + b;
+ *   grad();
+ *   return c.adj();
+ * });
+ *
+ * Doing so will not interfere with the process (or thread) AD tape.
+ */
+class ScopedChainableStack {
+  ChainableStack::AutodiffStackStorage local_stack_;
+
+  std::vector<ChainableStack::AutodiffStackStorage*> stack_queue_;
```

Comment:
As discussed in meeting, I don't like `std::vector` here. If we take away the stack, we have the limitation that a `ScopedChainableStack` 's execute function cannot be called recursively inside its own execute, but that is not a huge limitation.

Reply:
Sure... will change that.